File: core/src/main/java/de/thm/arsnova/service/TextRenderingService.java
package de.thm.arsnova.service;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import de.thm.arsnova.config.properties.SystemProperties;
import de.thm.arsnova.model.TextRenderingOptions;

@Service
public class TextRenderingService {
  private static final String ENDPOINT = "/render";
  private static final String MULTIPLE_ENDPOINT = "/render-multiple";
  private static final Logger logger = LoggerFactory.getLogger(TextRenderingService.class);

  private final boolean enabled;
  private final WebClient webClient;

  public TextRenderingService(final SystemProperties systemProperties) {
    this.enabled = systemProperties.getFormattingService().isEnabled();
    final String host = systemProperties.getFormattingService().getHostUrl();
    this.webClient = WebClient.create(host);
  }

  @Cacheable(value = "rendered-texts", unless = "#result == null")
  public String renderText(final String unrenderedText, final TextRenderingOptions textRenderingOptions)
      throws IOException {
    if (!enabled || unrenderedText == null || unrenderedText.isBlank()) {
      return null;
    }

    final RenderingRequestEntity requestEntity = new RenderingRequestEntity();
    requestEntity.setText(unrenderedText);
    requestEntity.setOptions(textRenderingOptions);
    logger.trace("Sending text to formatting service: {}", unrenderedText.hashCode());
    final RenderingResponseEntity responseEntity = webClient.post()
        .uri(ENDPOINT)
        .contentType(MediaType.APPLICATION_JSON)
        .accept(MediaType.APPLICATION_JSON)
        .bodyValue(requestEntity)
        .retrieve()
        .bodyToMono(RenderingResponseEntity.class)
        .doOnError(e -> logger.error("Text rendering request failed.", e))
        .onErrorResume(e -> Mono.just(new RenderingResponseEntity()))
        .block();
    return responseEntity.html;
  }

  @Cacheable(value = "rendered-texts", unless = "#result == null")
  public List<String> renderTexts(final List<String> unrenderedTexts, final TextRenderingOptions textRenderingOptions)
      throws IOException {
    if (!enabled || unrenderedTexts == null || unrenderedTexts.isEmpty()) {
      return null;
    }

    final List<RenderingRequestEntity> requestEntities = unrenderedTexts.stream().map(text -> {
      final RenderingRequestEntity requestEntity = new RenderingRequestEntity();
      requestEntity.setText(text);
      requestEntity.setOptions(textRenderingOptions);
      return requestEntity;
    }).collect(Collectors.toList());

    logger.trace("Sending texts to formatting service: {}", unrenderedTexts.hashCode());
    return webClient.post()
        .uri(MULTIPLE_ENDPOINT)
        .contentType(MediaType.APPLICATION_JSON)
        .accept(MediaType.APPLICATION_JSON)
        .bodyValue(requestEntities)
        .retrieve()
        .bodyToFlux(RenderingResponseEntity.class)
        .doOnError(e -> logger.error("Text rendering request failed.", e))
        .onErrorResume(e -> Flux.fromIterable(Collections.emptyList()))
        .map(e -> e.html)
        .collectList()
        .block();
  }

  private static class RenderingRequestEntity {
    private String text;
    private TextRenderingOptions options;

    public String getText() {
      return text;
    }

    public void setText(final String text) {
      this.text = text;
    }

    public TextRenderingOptions getOptions() {
      return options;
    }

    public void setOptions(final TextRenderingOptions options) {
      this.options = options;
    }
  }

  private static class RenderingResponseEntity {
    private String html;

    public String getHtml() {
      return html;
    }

    public void setHtml(final String html) {
      this.html = html;
    }
  }
}

File: core4/src/main/kotlin/net/particify/arsnova/core4/common/TextRenderingService.kt
/* Copyright 2025 Particify GmbH
 * SPDX-License-Identifier: MIT
 */
package net.particify.arsnova.core4.common

import com.fasterxml.jackson.annotation.JsonProperty
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClient

@Service
class TextRenderingService(properties: TextRenderingProperties) {
  companion object {
    val logger: Logger = LoggerFactory.getLogger(this::class.java)
  }

  private val client =
      RestClient.builder()
          .baseUrl(properties.serviceUrl)
          .defaultHeaders {
            it.accept = listOf(MediaType.APPLICATION_JSON)
            it.contentType = MediaType.APPLICATION_JSON
          }
          .build()

  fun renderText(
      unrenderedText: String?,
      textRenderingOptions: TextRenderingOptions = TextRenderingOptions()
  ): String? {
    if (unrenderedText.isNullOrBlank()) {
      return null
    }

    val requestEntity = RenderingRequestEntity(unrenderedText, textRenderingOptions)
    logger.trace("Sending text to rendering service: {}", unrenderedText.hashCode())
    val responseEntity: RenderingResponseEntity? =
        client
            .post()
            .uri("/render")
            .body(requestEntity)
            .retrieve()
            .body(RenderingResponseEntity::class.java)
    return responseEntity?.html
  }

  private data class RenderingRequestEntity(val text: String, val options: TextRenderingOptions)

  private data class RenderingResponseEntity(var html: String)

  data class TextRenderingOptions(
      @field:JsonProperty("linebreaks") val linebreaksEnabled: Boolean = true,
      @field:JsonProperty("markdown") val markdownEnabled: Boolean = true,
      @field:JsonProperty("latex") val latexEnabled: Boolean = true,
      @field:JsonProperty val markdownFeatureset: MarkdownFeatureset? = MarkdownFeatureset.SIMPLE
  ) {
    enum class MarkdownFeatureset {
      MINIMUM,
      SIMPLE,
      EXTENDED
    }
  }
}
